File: convert-sorted-array-to-binary-search-tree/convert-sorted-array-to-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode sortedArrayToBST(int[] nums) {
     if(nums.length==0) return null;
        return constructBinarySearchTree(nums,0,nums.length-1);
    }
    private TreeNode constructBinarySearchTree(int[] nums,int left,int right)
    {
        if(left>right) return null;
        int mid=left+(right-left)/2;
        TreeNode newNode=new TreeNode(nums[mid]);
        newNode.left=constructBinarySearchTree(nums,left,mid-1);
        newNode.right=constructBinarySearchTree(nums,mid+1,right);
        return newNode;
    }
}

File: convert-sorted-array-to-binary-search-tree/convert-sorted-array-to-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        if(nums.size()==0) return NULL;
        return constructBinaryTree(nums, 0, nums.size()-1);
    }
    TreeNode* constructBinaryTree(vector<int>& nums, int left, int right)
    {
        if(left>right) return NULL;
        int midpoint=left+(right-left)/2;
        TreeNode* newNode=new TreeNode(nums[midpoint]);
        newNode->left=constructBinaryTree(nums,left,midpoint-1);
        newNode->right=constructBinaryTree(nums,midpoint+1,right);
        return newNode;
    }
};